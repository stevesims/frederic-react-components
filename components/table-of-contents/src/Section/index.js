import React, { Component } from 'react';
import PropTypes from 'prop-types';

import Context from '../Context';

export class Entry extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    id: PropTypes.string.isRequired,
    title: PropTypes.node.isRequired,
    toc: PropTypes.shape({
      addContent: PropTypes.func.isRequired,
      removeContent: PropTypes.func.isRequired,
    }),
  }

  componentDidMount() {
    if(this.props.toc) {
      this.props.toc.addContent(this.props.id, this.props.title);
    }
  }

  componentWillUnmount() {
    if(this.props.toc) {
      this.props.toc.removeContent(this.props.id);
    }
  }

  render() {
    return <div id={this.props.id}>{this.props.children}</div>;
  }
}

const Section = ({standalone, ...props}) => {
  return standalone ?
    <Entry {...props} /> :
    <Context.Consumer>
      {toc => <Entry {...props} toc={toc} />}
    </Context.Consumer>;
};

Section.propTypes = {
  children: PropTypes.node.isRequired,
  id: PropTypes.string.isRequired,
  title: PropTypes.node.isRequired,
  standalone: PropTypes.bool,
};

export default Section;
