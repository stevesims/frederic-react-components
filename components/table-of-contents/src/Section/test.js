import React from 'react';
import { mount, shallow } from 'enzyme';

import Section, { Entry } from '.';

let wrapper;

describe('TableOfContents Section', () => {
  it('renders standalone', () => {
    wrapper = mount(
      <Section id="id" title="Title">Section text here</Section>,
    );
    expect(wrapper.exists()).toBe(true);
  });
});

describe('TableOfContents Section Entry', () => {
  describe('works with Context API', () => {
    const addContent = jest.fn();
    const removeContent = jest.fn();

    it('renders without crashing', () => {
      wrapper = shallow(
        <Entry id="id" title="Title" toc={{ addContent, removeContent }}>
          Section text here
        </Entry>,
      );
      expect(wrapper.exists()).toBe(true);
    });

    it('has called addContent in context', () => {
      expect(addContent.mock.calls.length).toBe(1);
    });

    it('has not yet called removeContent in context', () => {
      expect(removeContent.mock.calls.length).toBe(0);
    });

    it('calls removeContent on remove/unmount', () => {
      wrapper.unmount();
      expect(removeContent.mock.calls.length).toBe(1);
    });

    it('matches snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('works without Context API', () => {
    it('renders without crashing', () => {
      wrapper = mount(
        <Entry id="id" title="Title">
        Section text here
        </Entry>,
      );
      expect(wrapper.exists()).toBe(true);
    });

    it('does not find toc prop', () => {
      expect(wrapper.prop('toc')).toBeUndefined();
    });

    it('unmounts without issue', () => {
      wrapper.unmount();
    });

    it('matches snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
