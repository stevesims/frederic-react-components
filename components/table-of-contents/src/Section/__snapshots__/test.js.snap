// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableOfContents Section Entry works with Context API matches snapshot 1`] = `""`;

exports[`TableOfContents Section Entry works without Context API matches snapshot 1`] = `null`;
